flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/DataStreamRetractionRules.scala

```diff
@@ -0,0 +1,256 @@
+  * message, and update modifications as a pair of accumulate and retraction messages.
+  *
+  */
+object DataStreamRetractionRules {
+
+  /**
+    * Rule instance that assigns default retraction to [[DataStreamRel]] nodes.
+    */
+  val DEFAULT_RETRACTION_INSTANCE = new AssignDefaultRetractionRule()
+
+  /**
+    * Rule instance that checks if [[DataStreamRel]] nodes need to ship updates as retractions.
+    */
+  val UPDATES_AS_RETRACTION_INSTANCE = new SetUpdatesAsRetractionRule()
+
+  /**
+    * Rule instance that assigns the [[AccMode]] to [[DataStreamRel]] nodes.
+    */
+  val ACCMODE_INSTANCE = new SetAccModeRule()
+
+  /**
+    * Get all children RelNodes of a RelNode.
+    *
+    * @param parent The parent RelNode
+    * @return All child nodes
+    */
+  def getChildRelNodes(parent: RelNode): Seq[RelNode] = {
+    parent.getInputs.asScala.map(_.asInstanceOf[HepRelVertex].getCurrentRel)
+  }
+
+  /**
+    * Checks if a [[RelNode]] ships updates as retractions.
+    *
+    * @param node The node to check.
+    * @return True if the node ships updates as retractions, false otherwise.
+    */
+  def sendsUpdatesAsRetraction(node: RelNode): Boolean = {
+    val retractionTrait = node.getTraitSet.getTrait(UpdateAsRetractionTraitDef.INSTANCE)
+    retractionTrait != null && retractionTrait.sendsUpdatesAsRetractions
+  }
+
+  /**
+    * Rule that assigns the default retraction information to [[DataStreamRel]] nodes.
+    * The default is to not publish updates as retraction messages and [[AccMode.Acc]].
+    */
+  class AssignDefaultRetractionRule extends RelOptRule(
+    operand(
+      classOf[DataStreamRel], none()),
+    "AssignDefaultRetractionRule") {
+
+    override def onMatch(call: RelOptRuleCall): Unit = {
+      val rel = call.rel(0).asInstanceOf[DataStreamRel]
+      val traits = rel.getTraitSet
+
+      val traitsWithUpdateAsRetrac = if (!traits.contains(UpdateAsRetractionTraitDef.INSTANCE)) {
+        traits.plus(UpdateAsRetractionTrait.DEFAULT)
+      } else {
+        traits
+      }
+      val traitsWithAccMode =
+        if (!traitsWithUpdateAsRetrac.contains(AccModeTraitDef.INSTANCE)) {
+          traitsWithUpdateAsRetrac.plus(AccModeTrait.DEFAULT)
+      } else {
+        traitsWithUpdateAsRetrac
+      }
+
+      if (traits != traitsWithAccMode) {
+        call.transformTo(rel.copy(traitsWithAccMode, rel.getInputs))
+      }
+    }
+  }
+
+  /**
+    * Rule that annotates all [[DataStreamRel]] nodes that need to sent out update and delete
+    * changes as retraction messages.
+    */
+  class SetUpdatesAsRetractionRule extends RelOptRule(
+    operand(
+      classOf[DataStreamRel], none()),
+    "SetUpdatesAsRetractionRule") {
+
+    /**
+      * Checks if a [[RelNode]] requires that update and delete changes are sent with retraction
+      * messages.
+      */
+    def needsUpdatesAsRetraction(node: RelNode): Boolean = {
+      node match {
+          // should only be DataStreamGroupAggregate?
```

Comment:
Oh, sorry. That was just a comment for myself when I reviewed the PR. 
Should have removed that before committing and opening the PR. 
Sorry for the confusion.